### ISSUE TITLE
pyproject.toml: False positive: colorama

### ISSUE THREAD

**Issue by pawamoy** [tweag/FawltyDeps#307]
**Describe the bug**
Hey, nice project! I'm running this in one of my own, and `fawltydeps` says `colorama` is not declared, while it is. It also says it is unused (??), while it is as well.

**To Reproduce**
```bash
git clone https://github.com/pawamoy/griffe
cd griffe
fawltydeps src --deps pyproject.toml
```

Output:
```console
% fawltydeps src --deps pyproject.toml --detailed

These imports appear to be undeclared dependencies:
- 'colorama' imported at:
    src/griffe/cli.py:26
    src/griffe/diff.py:9
- 'typing_extensions' imported at:
    src/griffe/docstrings/dataclasses.py:15
    src/griffe/docstrings/google.py:42
    src/griffe/docstrings/numpy.py:79

These dependencies appear to be unused (i.e. not imported):
- 'aiofiles' declared in:
    pyproject.toml
- 'colorama' declared in:
    pyproject.toml
```

```console
% git grep colorama
pyproject.toml:    "colorama>=0.4",
src/griffe/cli.py:import colorama
src/griffe/cli.py:    colorama.deinit()
src/griffe/cli.py:    colorama.init()
src/griffe/diff.py:from colorama import Fore, Style
```

**Expected behavior**
`colorama` properly detected as used and declared.

**Environment**
 - OS name + version: 6.2.10-arch1-1
 - Version of the code: 0.8.0

**Comment by mknorps** [comment]
Hi @pawamoy !

Thank you for submitting the issue. 

I see you are using ArchLinux. Unfortunately, I cannot reproduce this problem on Ubuntu, and in our CI we test on Ubuntu-22.04 , which is why we might have missed the underlying problem.
The output on Ubuntu is, as expected:
```
fawltydeps src --deps pyproject.toml --detailed    

These imports appear to be undeclared dependencies:
- 'typing_extensions' imported at:
    src/griffe/docstrings/dataclasses.py:15
    src/griffe/docstrings/google.py:42
    src/griffe/docstrings/numpy.py:79

These dependencies appear to be unused (i.e. not imported):
- 'aiofiles' declared in:
    pyproject.toml
```

Could you run fawltydeps with verbose option so we can compare logs of running on those two systems?
```
fawltydeps src --deps pyproject.toml --detailed -vv
```

Could you also provide more details about the Python version and the environment (virtual env, nix-shell, sth else?) that you run  fawltydeps in? The dependency-to-import matching mechanism uses a virtual environment.

**Comment by jherland** [comment]
I've tried to dig into this issue, and although I don't have all the answers yet, I think the core of the issue is `__pypackages__`/PEP58: This way of installing Python packages is currently not supported/known to FawltyDeps, and when we resolve packages in the above logs, the `__pypackages__` directory is not involved at all.

Note that PEP582 [has been rejected](https://discuss.python.org/t/pep-582-python-local-packages-directory/963/430), and, AFAICS, PDM is now defaulting to `venv`-based dependency management instead (at least I had to configure my PDM to use PEP582). That said, since PEP582 is already being used "in the wild", we should consider supporting it if the burden is not too great.

Anyway, what happens here is that (since `--pyenv` is not used) we use the Python environment in which `fawltydeps` is installed. This is the domain of `pipx` which (in my experiments) ends up executing inside the virtualenv at `$HOME/.local/pipx/venvs/fawltydeps`. This virtualenv does not have colorama installed (in my experiments), which means it either finds colorama elsewhere in `sys.path`, or it does not find it at all (faling back to the identity mapping as a last resort).

- In your case, @pawamoy, it seems colorama is found somewhere in `sys.path`, but that the `importlib_metadata` version installed with fawltydeps is not able to establish any imports provided by this version. This is probably either an artifact of how/where colorama was installed, or a bug in `importlib_metadata`.
- In @mknorps's case, assuming she ran fawltydeps via `pipx`, is seems colorama was found _and_ `importlib_metadata` was able to deduce the correct mapping, which produced the right result in the end, but _not_ as a result of looking at the "correct" packages.
- In my case, I ran fawltydeps via `pipx` and got:
    ```
    INFO:fawltydeps.packages:'colorama' was not resolved. Assuming it can be imported as 'colorama'.
    DEBUG:fawltydeps.packages:  Resolved {'colorama': Package(package_name='colorama', mappings={<DependenciesMapping.IDENTITY: 'identity'>: {'colorama'}}, import_names={'colorama'}), ...} with <fawltydeps.packages.IdentityMapping object at 0x7fc5f20bdff0>
    ```
    In other words, no colorama was found anywhere, and we fell back to the identity mapping, assuming package `colorama` provides import `colorama` (which happens to work well in this case),

But as I said above, the core issue is our lack of PEP582 support. I'll post a PR soon trying to address that.

It's also interesting to pursue why `importlib_metadata` was not able to map colorama correctly in your case, @pawamoy: Can you post some more information to help track this down?
- The packages installed in your FawltyDeps virtualenv, e.g. the output of:
    ```
    ls -al $HOME/.local/pipx/venvs/fawltydeps/lib/python*/site-packages/
    ```
- The location where it found colorama in the end, e.g.:
    ```
    $HOME/.local/pipx/venvs/fawltydeps/bin/python -c "import colorama; print(colorama.__file__)"
    ```
- A directory listing of the colorama package metadata from that location, e.g.:
    ```
    ls -al $(dirname $($HOME/.local/pipx/venvs/fawltydeps/bin/python -c "import colorama; print(colorama.__file__)"))/../colorama*
    ```

**Comment by pawamoy** [comment]
Thank you so much for the thorough investigation and report @jherland.

```console
% ls -al $HOME/.local/pipx/venvs/fawltydeps/lib/python*/site-packages/
total 276
drwxr-xr-x 21 pawamoy users  4096 Apr 14 07:37 .
drwxr-xr-x  3 pawamoy users  4096 Apr 14 07:37 ..
drwxr-xr-x  3 pawamoy users  4096 Apr 14 07:37 fawltydeps
drwxr-xr-x  2 pawamoy users  4096 Apr 14 07:37 fawltydeps-0.8.0.dist-info
drwxr-xr-x  3 pawamoy users  4096 Apr 14 07:37 importlib_metadata
drwxr-xr-x  2 pawamoy users  4096 Apr 14 07:37 importlib_metadata-6.3.0.dist-info
drwxr-xr-x  7 pawamoy users  4096 Apr 14 07:37 isort
drwxr-xr-x  2 pawamoy users  4096 Apr 14 07:37 isort-5.11.5.dist-info
drwxr-xr-x  3 pawamoy users  4096 Apr 14 07:37 packaging
drwxr-xr-x  2 pawamoy users  4096 Apr 14 07:37 packaging-23.1.dist-info
-rw-r--r--  1 pawamoy users  5092 Apr 14 07:37 packaging_legacy_version.py
drwxr-xr-x  2 pawamoy users  4096 Apr 14 07:37 pip_requirements_parser-32.0.1.dist-info
-rw-r--r--  1 pawamoy users 96343 Apr 14 07:37 pip_requirements_parser.py
-rw-r--r--  1 pawamoy users    62 Apr 14 07:37 pipx_shared.pth
drwxr-xr-x  2 pawamoy users  4096 Apr 14 07:37 __pycache__
drwxr-xr-x  3 pawamoy users  4096 Apr 14 07:37 pydantic
drwxr-xr-x  2 pawamoy users  4096 Apr 14 07:37 pydantic-1.10.7.dist-info
drwxr-xr-x  4 pawamoy users  4096 Apr 14 07:37 pyparsing
drwxr-xr-x  2 pawamoy users  4096 Apr 14 07:37 pyparsing-3.0.9.dist-info
drwxr-xr-x  3 pawamoy users  4096 Apr 14 07:37 tomli
drwxr-xr-x  2 pawamoy users  4096 Apr 14 07:37 tomli-2.0.1.dist-info
drwxr-xr-x  2 pawamoy users  4096 Apr 14 07:37 typing_extensions-4.5.0.dist-info
-rw-r--r--  1 pawamoy users 84065 Apr 14 07:37 typing_extensions.py
drwxr-xr-x  3 pawamoy users  4096 Apr 14 07:37 zipp
drwxr-xr-x  2 pawamoy users  4096 Apr 14 07:37 zipp-3.15.0.dist-info

% $HOME/.local/pipx/venvs/fawltydeps/bin/python -c "import colorama; print(colorama.__file__)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'colorama'
```

It seems `colorama` isn't found in the `fawltydeps` pipx venv :thinking: 

---

That's great if you decide to support PEP 582. PDM's author has decided to keep support (for how long, I don't know) because it's an awesome feature. I saw that it was rejected but can't stop using it: it's so good :sweat_smile:

**Comment by jherland** [comment]
> Note that the Python interpreter of the `fawltydeps` venv is able to find `colorama` because I've [enabled PEP 582 globally](https://pdm.fming.dev/latest/usage/pep582/#enable-pep-582-globally) and I run this command in my `griffe` project root.

Aha, this points towards some issue between the colorama in `__pypackages__` and `importlib_metadata`'s ability to extract correct information from it. Unfortunately, I'm unable to reproduce on my end. Running with PR #312 applied (to get more debug info as well as pointing directly at `__pypackates__`:

```sh
$ fawltydeps src --deps pyproject.toml --pyenv __pypackages__

These imports appear to be undeclared dependencies:
- 'typing_extensions'

For a more verbose report re-run with the `--detailed` option.
$ fawltydeps src --deps pyproject.toml --pyenv __pypackages__ -vv 2>&1 | grep colorama
[...]
DEBUG:fawltydeps.packages:Trying to resolve {'colorama', 'cached-property'} with <fawltydeps.packages.LocalPackageResolver object at 0x7fe700c2b0a0>
DEBUG:fawltydeps.packages:Found colorama 0.4.6 under __pypackages__/3.10/lib
DEBUG:fawltydeps.packages:  Resolved {'colorama': Package(package_name='colorama', mappings={<DependenciesMapping.LOCAL_ENV: 'local_env'>: {'colorama'}}, import_names={'colorama'})} with <fawltydeps.packages.LocalPackageResolver object at 0x7fe700c2b0a0>
```

So in my case, I do get the expected `colorama -> colorama` mapping when pointing `importlib_metadata` at `__pypackages__/3.10/lib`.

Here are the steps that fawltydeps follow in excruciating details:
```sh
$ $HOME/.local/pipx/venvs/fawltydeps/bin/python
Python 3.10.9 (main, Dec  6 2022, 18:44:57) [GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import importlib_metadata as im
>>> context = im.DistributionFinder.Context(path=["__pypackages__/3.10/lib"])
>>> dists = list(im.MetadataPathFinder().find_distributions(context))
>>> [dist.name for dist in dists]
['Jinja2', 'mergedeep', 'typing_extensions', 'dparse', 'black', 'ruamel.yaml', 'Pygments', 'click', 'pyyaml_env_tag', 'six', 'colorama', 'charset-normalizer', 'ruff', 'mkdocs-literate-nav', 'python-dateutil', 'mypy', 'types-Markdown', 'mkdocs-section-index', 'MarkupSafe', 'toml', 'ruamel.yaml.clib', 'packaging', 'types-toml', 'setuptools', 'watchdog', 'markdown-it-py', 'requests', 'mkdocs', 'griffe', 'mkdocs-coverage', 'certifi', 'failprint', 'mkdocs-material', 'safety', 'tomli', 'mkdocstrings', 'mkdocs-autorefs', 'pathspec', 'ansimarkup', 'pymdown-extensions', 'mkdocstrings-python', 'idna', 'mypy-extensions', 'PyYAML', 'ptyprocess', 'mdurl', 'duty', 'Markdown', 'mkdocs-material-extensions', 'rich', 'regex', 'urllib3', 'editables', 'mkdocs-gen-files', 'markdown-callouts', 'platformdirs', 'ghp-import', 'markdown-exec']
>>> colorama = next(dist for dist in dists if dist.name == "colorama")
>>> colorama.files
[PackagePath('colorama/__init__.py'), PackagePath('colorama/ansi.py'), PackagePath('colorama/ansitowin32.py'), PackagePath('colorama/initialise.py'), PackagePath('colorama/win32.py'), PackagePath('colorama/winterm.py'), PackagePath('colorama/tests/__init__.py'), PackagePath('colorama/tests/ansi_test.py'), PackagePath('colorama/tests/ansitowin32_test.py'), PackagePath('colorama/tests/initialise_test.py'), PackagePath('colorama/tests/isatty_test.py'), PackagePath('colorama/tests/utils.py'), PackagePath('colorama/tests/winterm_test.py'), PackagePath('colorama-0.4.6.dist-info/METADATA'), PackagePath('colorama-0.4.6.dist-info/WHEEL'), PackagePath('colorama-0.4.6.dist-info/licenses/LICENSE.txt'), PackagePath('colorama-0.4.6.dist-info/RECORD')]
>>> im._top_level_declared(colorama)
[]
>>> list(im._top_level_inferred(colorama))
['colorama']
>>> imports = set(im._top_level_declared(colorama) or im._top_level_inferred(colorama))
>>> imports
{'colorama'}
```
It would be interesting to see where your environment behaves differently.

**Comment by pawamoy** [comment]
```pycon
Python 3.10.8 (main, Nov  5 2022, 15:38:20) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import importlib_metadata as im
>>> context = im.DistributionFinder.Context(path=["__pypackages__/3.10/lib"])
>>> dists = list(im.MetadataPathFinder().find_distributions(context))
>>> [dist.name for dist in dists]
['dparse', 'duty', 'iniconfig', 'markdown-callouts', 'mergedeep', 'typing-inspect', 'ghp-import', 'ansimarkup', 'regex', 'ptyprocess', 'Jinja2', 'git-changelog', 'mkdocs-gen-files', 'mkdocs-material', 'mkdocs-material-extensions', 'execnet', 'hypothesmith', 'aiofiles', 'safety', 'mkdocs-coverage', 'types-toml', 'mkdocs-section-index', 'urllib3', 'mypy-extensions', 'rich', 'Pygments', 'watchdog', 'setuptools', 'charset-normalizer', 'ruamel.yaml.clib', 'python-dateutil', 'libcst', 'mkdocs', 'mkdocstrings-python', 'hypothesis', 'pytest-xdist', 'pytest-randomly', 'packaging', 'pytest-cov', 'editables', 'Markdown', 'pyyaml_env_tag', 'typing_extensions', 'blacken-docs', 'mkdocs-autorefs', 'pathspec', 'mdurl', 'ruff', 'pyrsistent', 'semver', 'requests', 'ruamel.yaml', 'pytest', 'click', 'attrs', 'idna', 'failprint', 'six', 'platformdirs', 'tomli', 'black', 'markdown-exec', 'griffe', 'mypy', 'exceptiongroup', 'pymdown-extensions', 'dependenpy', 'MarkupSafe', 'toml', 'certifi', 'markdown-it-py', 'PyYAML', 'mkdocstrings', 'colorama', 'types-aiofiles', 'sortedcontainers', 'types-Markdown', 'pluggy', 'jsonschema', 'lark', 'coverage', 'mkdocs-literate-nav']
>>> colorama = next(dist for dist in dists if dist.name == "colorama")
>>> colorama.files
[PackagePath('colorama-0.4.6.dist-info/METADATA'), PackagePath('colorama-0.4.6.dist-info/WHEEL'), PackagePath('colorama-0.4.6.dist-info/licenses/LICENSE.txt'), PackagePath('colorama-0.4.6.dist-info/REFER_TO'), PackagePath('colorama-0.4.6.dist-info/RECORD'), PackagePath('colorama')]
>>> im._top_level_declared(colorama)
[]
>>> list(im._top_level_inferred(colorama))
[]
>>> imports = set(im._top_level_declared(colorama) or im._top_level_inferred(colorama))
>>> imports
set()
```

**Comment by jherland** [comment]
Thanks, this was helpful! Our `colorama.files` are different: Where I have these extra items:
```
PackagePath('colorama/__init__.py'),
PackagePath('colorama/ansi.py'),
PackagePath('colorama/ansitowin32.py'),
PackagePath('colorama/initialise.py'),
PackagePath('colorama/win32.py'),
PackagePath('colorama/winterm.py'),
PackagePath('colorama/tests/__init__.py'),
PackagePath('colorama/tests/ansi_test.py'),
PackagePath('colorama/tests/ansitowin32_test.py'),
PackagePath('colorama/tests/initialise_test.py'),
PackagePath('colorama/tests/isatty_test.py'),
PackagePath('colorama/tests/utils.py'),
PackagePath('colorama/tests/winterm_test.py'),
```
you only have these:
```
PackagePath('colorama-0.4.6.dist-info/REFER_TO'),
PackagePath('colorama')
```

The `_top_level_inferred()` function in `importlib_metadata` uses the Python files returned by `colorama.files` to deduce which import names are exposed by a package. For example, since I have `PackagePath('colorama/__init__.py')`, `importlib_metadata` can deduce that there is a Python package inside the `colorama/` directory, and therefore that `colorama` must be a name that can be imported. In your case, there are no Python files in `colorama.files`, so `importlib_metadata` deduces that there are no imports provided by this package.

To figure out why, I need to learn more about the files you have:
- Is `__pypackages__/3.10/lib/colorama` a symlink? If so, where does it point?
- What is the contents of `__pypackages__/3.10/lib/colorama-0.4.6.dist-info/REFER_TO`?

Googling around to learn about this file, it is barely mentioned in the very recent [PEP710](https://peps.python.org/pep-0710/) as a file that is created by PDM when installing wheels with the `installer` tool. Looking at the [relevant code in PDM](https://github.com/pdm-project/pdm/blob/d39a8e5b36c37093ea31e666d0e55fe21b38c16b/src/pdm/installers/installers.py#L197) it seems that there may indeed be symlinks involved (search for `use_symlink` in that file), and I would further speculate that `importlib_metadata` does not handle these correctly.

In other words, this might be a combination of PDM doing something weird (not necessarily incorrect, but certainly unexpected), and `importlib_metadata` not being prepared for this weirdness...

**Comment by pawamoy** [comment]
I'm amazed by how well you're able to synthesize your findings @jherland 🥲

I'm indeed using PDM's cache feature, which makes it so that packages in `__pypackages__` are actually symlinks to their cache versions, in `~/.cache/pdm/packages`.

It would make sense that `importlib.metadata` then doesn't handle symlinks that well.

**Comment by jherland** [comment]
Thanks, I was finally able to reproduce after `pdm config install.cache true`. Looking further into `importlib_metadata` (we're using the separate package rather than the stdlib `importlib.metadata` which lags behind in features), I can confirm what I suspected above:
- `importlib_metadata` uses the contents of `__pypackages__/3.10/lib/colorama-0.4.6.dist-info/RECORD` to produce `colorama.files`:
    ```
    colorama-0.4.6.dist-info/METADATA,sha256=e67SnrUMOym9sz_4TjF3vxvAV4T3aF7NyqRHHH3YEMw,17158
    colorama-0.4.6.dist-info/WHEEL,sha256=cdcF4Fbd0FPtw2EMIOwH-3rSOTUdTCeOSXRMD1iLUb8,105
    colorama-0.4.6.dist-info/licenses/LICENSE.txt,sha256=ysNcAmhuXQSlpxQL-zs25zrtSWZW6JEQLkKIhteTAxg,1491
    colorama-0.4.6.dist-info/REFER_TO,sha256=b2lWg3Osknh6HODITiPmZW55TlJ8okVt-8k6rf3kdpc,66
    colorama-0.4.6.dist-info/RECORD,,
    colorama,,
    ```
- Since `importlib_metadata` does not actually traverse any directories listed in `RECORD` (frankly, I suspect it does not even expect to find any directories or symlinks in `RECORD` at all) it never sees any Python files within the `colorama` subdir:
- I've created https://github.com/python/importlib_metadata/issues/452 and https://github.com/python/importlib_metadata/issues/453 to see what the importlib_metadata maintainer thinks about this issue. Until then, I suspect a workaround is to use `pdm config install.cache false`.

**Comment by pawamoy** [comment]
Thanks a lot for getting to such length in fixing this issue, this is greatly appreciated :rocket:

**Comment by jherland** [comment]
Closing for now. A workaround exists (disable `install.cache` in PDM). We will follow the importlib_metadata issue separately.
